cf-update-circl.py

```diff
@@ -0,0 +1,195 @@
+#!/usr/bin/env python3
+#
+# cf-update-circl.py updates the CIRCL package in src/circl.
+import os
+import sys
+import tempfile
+import subprocess
+
+# Name of the CIRCL package as it will appear in the standard library.
+circl_name = 'circl'
+
+# The git repository form which CIRCL will be cloned.
+circl_repo = 'https://github.com/cloudflare/circl'
+
+# The git branch to checkout.
+circl_branch = 'master'
+
+# The set of CIRCL packages that the standard library depends on.
+circl_pkg_dir = [
+  'circl',
+  'circl/ecc/goldilocks',
+  'circl/dh/sidh',
+  'circl/dh/sidh/internal/common',
+  'circl/dh/sidh/internal/p434',
+  'circl/dh/sidh/internal/p503',
+  'circl/dh/sidh/internal/p751',
+  'circl/internal/conv',
+  'circl/internal/sha3',
+  'circl/internal/test',
+  'circl/math',
+  'circl/math/fp25519',
+  'circl/math/fp448',
+  'circl/math/mlsbset',
+  'circl/pki',
+  'circl/sign',
+  'circl/sign/dilithium/internal/common',
+  'circl/sign/dilithium/internal/common/params',
+  'circl/sign/dilithium/mode3',
+  'circl/sign/dilithium/mode3/internal',
+  'circl/sign/dilithium/mode4',
+  'circl/sign/dilithium/mode4/internal',
+  'circl/sign/ed25519',
+  'circl/sign/ed448',
+  'circl/sign/eddilithium3',
+  'circl/sign/eddilithium4',
+  'circl/sign/schemes',
+  'circl/simd/keccakf1600',
+]
+
+# The directories of test data required for CIRCL unit tests.
+circl_testdata_dir = [
+  'circl/dh/sidh/testdata',
+  'circl/internal/sha3/testdata',
+  'circl/sign/ed25519/testdata',
+  'circl/sign/ed448/testdata',
+]
+
+# Miscellaneous files to copy from CIRCL.
+circl_misc_file = [
+  'circl/LICENSE',
+]
+
+src = os.path.join(os.path.dirname(os.path.abspath(sys.argv[0])), 'src')
+src_circl = os.path.join(src, circl_name)
+
+# Returns true if fn is a source file for CIRCL. Currently this includes Go
+# sources code, C header files, and ASM source code.
+def is_pkg_file(fn):
+    ok = False
+    for suffix in ['go', 'h', 's']:
+      ok |= fn.endswith('.' + suffix)
+    return ok
+
+# Adds header to the top of file fn and, if fn is a Go source file, replaces
+# imports with the corresponding package in the standard library.
+def handle_pkg_file(fn, header):
+    with open(fn, 'rb') as f:
+        s = f.read()
+
+    # Fix Go imports.
+    if fn.endswith('.go'):
+        s = s.replace(
+            b'github.com/cloudflare/circl',
+            b'circl',
+        ).replace(
+            b'golang.org/x/sys/cpu',
+            b'internal/cpu',
+        )
+
+    with open(fn, 'wb') as f:
+        f.write(header)
+        f.write(s)
+
+pkg_file_header = '''
+// THIS CODE WAS GENERATED BY cf-update-circl.py.
+//
+// It was checked out from "{0}"
+// on branch "{1}". Latest commit:
+//'''
```

Comment:
Let's not add it to every individual file. This info can be read from the git history for the Go repo. If it is really important to have it in the source tree, I would suggest creating one file with this metadata, e.g. `src/circle/about.txt`

Reply:
Just curious, what do you see as the downside of adding it to each file? Big diffs?

Reply:
(Done.)

Reply:
Big diffs to a lesser extent, I was more thinking about comparing two files or directories (src/circl/ vs circl upstream).